package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,335 @@
+        selection : str
+            An MDAnalysis selection string. Defaults to `None`.
+        msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Desired dimensions to be included in the MSD. Defaults to 'xyz'.
+        fft : bool
+            Use a fast FFT based algorithm for computation of the MSD.
+            Otherwise, use the naieve or "simple" algorithm. Defaults to `True`
+
+        """
+        #args
+        self.u = u
+        self.selection = selection
+        self.msd_type = msd_type
+        self.fft = fft
+        
+        #local
+        self._dim = None
+        self._position_array = None
+        
+        #indexing
+        self.N_frames = len(self.u.trajectory)
+        self.N_particles = 0
+
+        #result
+        self.dim_fac = 0
+        self.timeseries = None
+        self.msd_per_particle = None
+
+        #prep
+        self._prepare()
+
+            
+    def _prepare(self):
+        self._parse_msd_type()
+        self._select_reference_positions()
+        
+    def _parse_msd_type(self):
+        r""" Sets up the desired dimensionality of the MSD.
+        
+        Parameters
+        ----------
+        self.msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Dimensions to be included in the MSD.
+
+        Returns
+        -------
+        self._dim : list
+            Array-like used to slice the positions to obtain desired dimensionality
+        self.dim_fac : float
+            Dimension factor :math:`d` of the MSD.
+
+        """
+        keys = {'x':0, 'y':1, 'z':2}
+        dim = []
+
+        self.msd_type = self.msd_type.lower()
+        print(self.msd_type)
+        try:
+            for tok in self.msd_type:
+                vals = (keys.pop(tok))
+                dim.append(vals)
+        except KeyError:
+            raise ValueError('invalid msd_type specified, please specify one of xyz, xy, xz, yz, x, y, z ')
+        
+        dim.sort()
+        self._dim = dim
+        self.dim_fac = float(len(dim))
+
+    def _select_reference_positions(self):
+        r""" Constructs array of positions for MSD calculation.
+        
+        Parameters
+        ----------
+        self.u : :class:`Universe`
+            MDAnalysis Universe
+        self.selection : str
+            MDAnalysis selection string
+        
+        Returns
+        -------
+        self._position_array : :class:`np.ndarray`
+            Array of particle positions with respect to time shape = (N_frames, N_particles, 3)
+        self.N_particles : float
+            Number of particles used in the MSD calculation
+
+        """
+        #avoid memory transfer by iterating 
+        idxs = self.u.select_atoms(self.selection).ix
+        position_array_dummy = np.zeros([len(self.u.trajectory), len(self.u.atoms), 3])
+        for idx,ts in enumerate(self.u.trajectory):
+           position_array_dummy[idx,:,:] = self.u.atoms.positions
+        self._position_array = position_array_dummy[:,idxs, :]
```

Comment:
You could do this directly with 
```
ag = self.u.select_atoms(self.selection)
self._position_array = [len(self.u.trajectory), len(ag), 3]
```
although do you need all 3 XYZ coordinates if you have only 1 dimension selected by the user?

If you subclassed AnalysisBase, it would look like

```
self._position_array = [self.n_frames, len(ag), 3]
```